maxtest/keyexpr.ms

```diff
@@ -0,0 +1,175 @@
+fileIn @"model.ms"
+
+
+/*
+ *  KeyExpressionMatcher is a module that can select strings (in
+ *  this case testnames) that match a given keyword expression.
+ *
+ *  Returns matching tests as objects the runner can use to execute the tests.
+ */
+struct KeyExpressionMatcher
+(
+
+    private _ignoredTokens =  #("and", "or", "not", ""),
+
+    public
+    function getCombinedTestName testCase =
+    (
+        /*
+         *  <String>getCombinedTestName <MAXTestCase>testCase
+         *  Returns a combination of file, struct and function name.
+         */
+        local testSuite = testCase.testSuite
+        local suiteName = testSuite.structName
+        local fileName = getFileNameFile testSuite.filePath
+        local testName = testCase.testFnName
+        local combinedName = fileName + "." + suiteName + "." + testName
+        combinedName
+    ),
+
+
+    public
+    function getTokens expr =
+    (
+        /*
+         *  <Array of String>getTokens <String>expr
+         *  Gets a keyword expression and returns all valid keywords.
+         *  Note: Is not case-sensitive.
+         */
+        local space = " "
+        local replaceThese = #("(", ")")
+
+        -- Replace all parentheses by spaces and split by that.
+        expr = toLower expr
+        for replaceThis in replaceThese do (
+            expr = substituteString expr replaceThis space
+        )
+        local rawTokens = filterString expr space
+        local tokens = #()
+
+        -- Capture tokens that are neither empty nor reserved.
+        for rawToken in rawTokens do (
+            local skip = false
+            local token = rawToken
+            for ignore in this._ignoredTokens do (
+                if ignore == token do (
+                    skip = true
+                    continue
+                )
+            )
+            if not skip do (
+                appendIfUnique tokens token
+            )
+        )
+
+        tokens
+    ),
+
+
+    public
+    function substituteTokens expr tokenValuePairs = (
+        /*
+         *  <String>substituteTokens <String>expr <Array of Arrays>tokenValuePairs
+         *  Substitutes tokens in the expressions by their boolean values.
+         *
+         *  The input list contains pairs of #(<token>, <bool>) which is
+         *  used for substitution. Returns the complete substituted
+         *  expression, ready for evaluation.
+         *
+         * TODO: We need to protect ignored tokens from substitution
```

Comment:
:grey_exclamation: Should be doable by decomposing the expression into its tokens (protected and normal ones) and substituting only those that are not protected, than rebuild the expression from the results.